### ISSUE TITLE
TreeDataProvider filters only parents

### ISSUE THREAD

**Issue by Klaudeta** [vaadin/flow#5106]
When filtering a TreeGrid which uses TreeDataProvider if a child item pass the filter but any of its parent/grandparent does not than the child it not rendered at all on the grid because that parent in the change is not rendered.

Code to reproduce:

```
@Route("")
@PWA(name = "Project Base for Vaadin Flow", shortName = "Project Base")
public class MainView extends VerticalLayout {

    
    public MainView() {
        List<Bean> parents = new ArrayList<>();
        Map<Bean, List<Bean>> map = new HashMap<>();

        Bean b = new Bean("Europe");
        Bean s = new Bean("United States");
        parents.add(b); parents.add(s);
        map.put(b, Arrays.asList(new Bean("Albania", b), new Bean("Finland", b), new Bean("Italy", b)));
        map.put(s, Arrays.asList(new Bean("California", s),new Bean("Washington", s),new Bean("New York", s)));

        TreeGrid<Bean> grid = new TreeGrid<>();

        grid.addHierarchyColumn(Bean::getName).setHeader("Name");
        TreeData<Bean> treeData = new TreeData<>();
        treeData.addItems(parents, item -> map.get(item) != null ? map.get(item) : Collections.emptyList());
        TreeDataProvider<Bean> dp = new TreeDataProvider<>(treeData);
        grid.setDataProvider(dp);

        TextField filter = new TextField("Filter By");
        filter.addValueChangeListener(e -> {
            dp.setFilter(item -> {
                return item.getName().toLowerCase().contains(e.getValue());
            });
        });

        add(filter,grid);
    }

    public static class Bean{
        private String name;

        private Bean parent;

        public Bean(String name) {
            this.name = name;
        }

        public Bean(String name, Bean parent) {
            this.name = name;
            this.parent = parent;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public Bean getParent() {
            return parent;
        }

        public void setParent(Bean parent) {
            this.parent = parent;
        }
    }

}
```

PS: I'm posting this here as it is an issue with the TreeGridDataProvider rather than with the TreeGrid component itself.

**Comment by jm2rodriguez** [comment]
In my opinion, the filter behavior on a TreeDataProvider should be as follows.
- If the search term is found in a parent element then all its descendants must be displayed in the tree structure.
- If the search term is found in a child element then all its ancestors and siblings must be displayed in the tree structure.

The filter should therefore apply to a set of hierarchically related elements (parents, childs, siblings) and not to a single element independent of the hierarchical structure in which it is located.

**Comment by TatuLund** [comment]
Isn't this fixed by https://github.com/vaadin/flow/pull/7317 ?

**Comment by pleku** [comment]
>Isn't this fixed by #7317 ?

Yes, good catch.
